Windows hacking

hslatman opened the pull request.

brandonweeks committed:
x509ext: initial version of package

mjg59 committed:
Add TPM 2 application key support for Windows

There's currently no support for creating application keys on Windows systems. This patch transitions the Windows key type to specifically refer to attestation keys, and reuses the existing wrapped key support for application keys. This allows the creation of keys in the platform store, while still allowing said keys to be manipulated with existing TPM functionality rather than duplicating it.

brandonweeks committed:
Add QualifyingData paramater to KeyConfig

hslatman committed:
Merge remote-tracking branch 'bweeks/x509ext' into patches

hslatman committed:
Merge remote-tracking branch 'mjg59/windows_application_keys' into patches

hslatman committed:
Merge remote-tracking branch 'bweeks/acme-device-attest' into patches

hslatman committed:
Fix `QualifyingData` on Windows

hslatman committed:
Add signing support for keys generated on Windows

When generating a new key using a Windows TPM, a `wrappedKey20` was
returned, which couldn't be used for signing on Windows, as it's
backed by a `windowsTPM`. The `wrappedKey20` seems to be a type
specifically aimed at usage with a `wrappedTPM20`, which in turn
seems to be used on Linux and for testing, but not when instantiating
a TPM on Windows.

This commit adds the `newWindowsKey20` function, which returns
a key backed by a `windowsTPM`. The key is a `windowsAK20`,
now also conforming to the `key` interface, so that it can be used
for signing purposes.

hslatman committed:
Implement `blobs` for Windows keys

hslatman committed:
Add `loadKey` on Windows

hslatman committed:
Remove superfluous return

hslatman committed:
Add `Name` and `Prefix` options to `KeyConfig`

hslatman committed:
Disable Windows TPM check for `KeyConfig` properties

hslatman committed:
Add `Name` and `Prefix` to AK creation

hslatman committed:
Merge branch 'master' into herman/windows-hacking

hslatman committed:
Fix missing `ECDSA` curve when signing on Windows

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/smallstep/go-attestation?pullRequest=2) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/smallstep/go-attestation?pullRequest=2) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: hslatman<br/>:x: mjg59<br/>:x: brandonweeks<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/smallstep/go-attestation?pullRequest=2) it.</sub>

hslatman committed:
Add support for deleting keys

On Windows, when the key is managed by the OS, keys are stored on
the filesystem. When trying to create a key with the same name, this will
fail with the following error:

`NCryptCreatePersistedKey returned 8009000F: The operation completed successfully.`

This commit adds support for deleting these keys, so that a new
key can be created with the same name.

Have only tested this on Windows so far. My assumption is that for
keys created with `go-attestation` on Linux, the keys aren't persisted
"inside the TPM", so there's nothing to do when deleting them, except
for any keys managed externally.

hslatman committed:
Fix tests for TPM 1.2

hslatman committed:
Merge branch 'master' into herman/windows-hacking

hslatman committed:
Fix tests for TPM 1.2

hslatman committed:
Add `DeleteAK` method

hslatman committed:
Add `Blobs` method for `AK`

hslatman committed:
Add missing methods for TPM 1.2

hslatman committed:
Merge branch 'master' into herman/windows-hacking

hslatman committed:
Remove key prefix and some cleanup

hslatman committed:
Merge branch 'master' into herman/windows-hacking

hslatman committed:
Fix legacy TPM2 path for Windows key creation and loading
